### ISSUE TITLE
Scatter Fit

### ISSUE THREAD

**Issue by annayqho** [annayqho/AnnaCannon#26]
Currently we step through different possible scatter values, and for each value we perform a least squares fit to solve for the coefficients. In the first released version of the code, so we want to optimize for the scatter term somehow instead of doing a guess-and-check? @davidwhogg , @HWRix , @mkness

**Comment by davidwhogg** [comment]
Really?  I thought we used an optimizer!  Can you point me into the code + line number in your code, and the code + line number in @mkness 's code?

**Comment by mkness** [comment]
Currently for the scatter we input an array of scatter values to calculate the chi2 for these values and use a polynomial fit to return the minimum - I thought that still counted as "optimisation" although I didn't upgrade it to actually use the optimiser function (is still an open issue!). I haven't forgotten this is an open issue I am just slow... Line 673. I will make this a priority to try to upgrade this part of the code if necessary.

**Comment by davidwhogg** [comment]
Well, we should start by benchmarking an optimizer -- but we should only optimize in 1-d (the scatter direction); at each scatter we should solve the linear algebra.  Let's discuss before implementation.

**Comment by davidwhogg** [comment]
ps by "benchmark" I mean compare speed between what we do now and an optimizer; I think the optimizer will be fast but we should make sure.

**Comment by annayqho** [comment]
OK, good to know (my code is basically the same as Melissa's, lines below)
z = np.polyfit(ln_scatter_vals_short, chis_eval_short, 2)
fit_pder = np.polyder(z)
best_scatter = np.exp(np.roots(fit_pder)[0])

Perhaps the three of us could discuss this over Skype sometime? It would be good to get it completely straightened out before I code anything new.

**Comment by davidwhogg** [comment]
does np.polyder() take the first derivative?

**Comment by annayqho** [comment]
Yes, 1st order differentiation is the default, but can be changed with an additional argument.

**Comment by davidwhogg** [comment]
Okay then I agree with this code.  The only question is whether one-d optimization would be faster.  But I would downgrade this issue to "low priority" in the short term.
